domain/src/test/java/org/oppia/domain/util/StateRetrieverTest.kt

```diff
@@ -0,0 +1,280 @@
+package org.oppia.domain.util
+
+import android.app.Application
+import android.content.Context
+import androidx.test.core.app.ApplicationProvider
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import com.google.common.truth.Truth.assertThat
+import dagger.BindsInstance
+import dagger.Component
+import dagger.Module
+import dagger.Provides
+import javax.inject.Inject
+import javax.inject.Qualifier
+import javax.inject.Singleton
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.test.TestCoroutineDispatcher
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.oppia.app.model.AnswerGroup
+import org.oppia.app.model.InteractionObject
+import org.oppia.app.model.ListOfSetsOfHtmlStrings
+import org.oppia.app.model.RuleSpec
+import org.oppia.app.model.State
+import org.oppia.app.model.StringList
+import org.oppia.util.caching.CacheAssetsLocally
+import org.oppia.util.logging.EnableConsoleLog
+import org.oppia.util.logging.EnableFileLog
+import org.oppia.util.logging.GlobalLogLevel
+import org.oppia.util.logging.LogLevel
+import org.oppia.util.threading.BackgroundDispatcher
+import org.oppia.util.threading.BlockingDispatcher
+import org.robolectric.annotation.Config
+
+const val DRAG_DROP_TEST_EXPLORATION_NAME = "drag_and_drop_test_exploration.json"
+
+/** Tests for [StateRetriever]. */
+@RunWith(AndroidJUnit4::class)
+@Config(manifest = Config.NONE)
+class StateRetrieverTest {
+
+  @Inject
+  lateinit var stateRetriever: StateRetriever
+
+  @Inject
+  lateinit var jsonAssetRetriever: JsonAssetRetriever
+
+  @Before
+  fun setUp() {
+    setUpTestApplicationComponent()
+  }
+
+  @Test
+  fun testParseState_withDragAndDropInteraction_parsesRuleWithIsEqualToOrderingRuleSpec() {
+    val state = createStateFromJson("DragDropSortInput", DRAG_DROP_TEST_EXPLORATION_NAME)
+    val ruleSpecMap = state.interaction.answerGroupsList
+      .flatMap(AnswerGroup::getRuleSpecsList)
+      .associateBy(RuleSpec::getRuleType)
+    assertThat(ruleSpecMap).containsKey("IsEqualToOrdering")
+  }
+
+  @Test
+  fun testParseState_withDragAndDropInteraction_parsesRuleWithIsEqualToOrderingWithInputXAndValueAtX() {
+    val state = createStateFromJson("DragDropSortInput", DRAG_DROP_TEST_EXPLORATION_NAME)
+    val ruleSpecMap = state.interaction.answerGroupsList
```

Comment:
Maybe move this to a helper method. E.g.:

```kotlin
private fun lookUpRuleSpec(state, ruleSpecName) {
  return state.interaction.answerGroupsList
      .flatMap()
      ...
}
```

Reply:
done